nucliadb_vectors/src/data_point_provider/segment_manager.rs

```diff
@@ -0,0 +1,574 @@
+use nucliadb_core::fs_state::{self, FsResult, Version};
+use serde::{Deserialize, Serialize};
+
+use crate::data_point::{DpId, Journal};
+use crate::data_types::dtrie_ram::DTrie;
+use crate::data_types::DeleteLog;
+use crate::VectorR;
+
+type TxId = u64;
+type SegmentId = DpId;
+
+#[derive(Clone, Copy)]
+struct TimeSensitiveDLog<'a> {
+    dlog: &'a DTrie,
+    time: TxId,
+}
+impl<'a> DeleteLog for TimeSensitiveDLog<'a> {
+    fn is_deleted(&self, key: &[u8]) -> bool {
+        self.dlog.get(key).map(|t| t > self.time).unwrap_or_default()
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
+struct JournalTransaction {
+    txid: TxId,
+    operations: Vec<Operation>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug, Clone, Copy)]
+enum Operation {
+    AddSegment(SegmentId),
+    DeleteSegment(SegmentId),
+}
+
+#[derive(Serialize, Deserialize, Default, Clone)]
+struct State {
+    journal: Vec<JournalTransaction>,
+    delete_log: DTrie,
+    no_nodes: usize,
+}
+
+#[derive(Default)]
+pub struct Transaction {
+    operations: Vec<Operation>,
+    deleted_entries: Vec<String>,
+    no_nodes: usize,
+}
+
+impl Transaction {
+    pub fn add_segment(&mut self, dp_journal: Journal) {
+        self.operations.push(Operation::AddSegment(dp_journal.id()));
+        self.no_nodes += dp_journal.no_nodes();
+    }
+
+    pub fn replace_segments(&mut self, old: Vec<SegmentId>, new: SegmentId) {
+        for dpid in old {
+            self.operations.push(Operation::DeleteSegment(dpid));
+        }
+        self.operations.push(Operation::AddSegment(new));
+    }
+
+    pub fn delete_entry(&mut self, prefix: String) {
+        self.deleted_entries.push(prefix);
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.operations.is_empty() && self.deleted_entries.is_empty()
+    }
+}
+
+const STATE_FILE_EXTENSION: &str = "segstate";
+
+struct StateFile {
+    path: PathBuf,
+    file: File,
+}
+
+impl StateFile {
+    fn new(path: PathBuf) -> FsResult<Self> {
+        let file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(path.join(format!("{}.{STATE_FILE_EXTENSION}", SegmentId::new_v4())))?;
+        file.lock_exclusive()?;
+
+        Ok(StateFile {
+            path,
+            file,
+        })
+    }
+
+    fn write(&mut self, data: impl Serialize) -> FsResult<()> {
+        self.file.seek(std::io::SeekFrom::Start(0))?;
+        bincode::serialize_into(&mut self.file, &data)?;
+        self.file.flush()?;
+
+        Ok(())
+    }
+
+    fn try_read<D: for<'a> Deserialize<'a>>(path: &Path) -> VectorR<Option<D>> {
+        let Some(extension) = path.extension() else {
+            return Ok(None);
+        };
+        if extension != OsStr::new(STATE_FILE_EXTENSION) {
+            return Ok(None);
+        }
+        let file = File::open(path)?;
+        let is_locked = file.try_lock_exclusive().is_err();
+        if !is_locked {
+            std::fs::remove_file(path)?;
```

Comment:
Oh, yeah, good catch, thanks! I don't think we need two files, we can create and lock atomically using a rename (or `O_TMPFILE`). So the only change needed is in the `StateFile::new` to do:
```
let file = OpenOptions::open("random.tmp")?;
file.lock_exclusive()?;
fs::rename("random.tmp", "random.segstate")
```

Since rename in the same FS only changes the parent directory entry, and does not touch the inode of the file, it's fine to do it with an open file and even while holding locks.

Reply:
Good insight! Seems like a good approach to me!